relay-redis/src/pool.rs

```diff
@@ -0,0 +1,210 @@
+/// ensuring proper connection health through periodic PING checks. It supports both
+/// primary and replica nodes, with optional read-from-replicas functionality.
+#[derive(Debug)]
+pub struct CustomClusterManager {
+    manager: ClusterManager,
+    interval_counter: IntervalCounter,
+}
+
+impl CustomClusterManager {
+    /// Creates a new cluster manager for the specified Redis nodes.
+    ///
+    /// The manager will attempt to connect to each node in the provided list and
+    /// maintain connections to the Redis cluster.
+    pub fn new<T: IntoConnectionInfo>(
+        params: Vec<T>,
+        read_from_replicas: bool,
+        refresh_interval: usize,
+    ) -> RedisResult<Self> {
+        Ok(Self {
+            manager: ClusterManager::new(params, read_from_replicas)?,
+            interval_counter: IntervalCounter::new(refresh_interval),
+        })
+    }
+}
+
+impl Manager for CustomClusterManager {
+    type Type = ClusterConnection;
+    type Error = RedisError;
+
+    async fn create(&self) -> Result<ClusterConnection, RedisError> {
+        self.manager.create().await
+    }
+
+    async fn recycle(
+        &self,
+        conn: &mut ClusterConnection,
+        metrics: &Metrics,
+    ) -> RecycleResult<RedisError> {
+        // If the interval has been reached, we optimistically assume the connection is active
+        // without doing an actual `PING`.
+        if !self.interval_counter.is_reached() {
+            return Ok(());
+        }
+
+        self.manager.recycle(conn, metrics).await
+    }
+}
+
+impl From<Object<CustomClusterManager>> for CustomClusterConnection {
+    fn from(conn: Object<CustomClusterManager>) -> Self {
+        Self(conn)
+    }
+}
+
+/// A managed Redis connection that implements [`redis::aio::ConnectionLike`].
+///
+/// This type provides access to the underlying Redis connection while managing its lifecycle.
+/// It is designed to be used with connection pools to efficiently handle Redis operations.
+pub struct CustomSingleConnection(Object<CustomSingleManager>);
+
+impl redis::aio::ConnectionLike for CustomSingleConnection {
+    fn req_packed_command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisFuture<'a, Value> {
+        self.0.req_packed_command(cmd)
+    }
+
+    fn req_packed_commands<'a>(
+        &'a mut self,
+        cmd: &'a Pipeline,
+        offset: usize,
+        count: usize,
+    ) -> RedisFuture<'a, Vec<Value>> {
+        self.0.req_packed_commands(cmd, offset, count)
+    }
+
+    fn get_db(&self) -> i64 {
+        self.0.get_db()
+    }
+}
+
+/// Manages single Redis instance connections and their lifecycle.
+///
+/// This manager handles the creation and recycling of Redis connections,
+/// ensuring proper connection health through periodic PING checks. It supports
+/// multiplexed connections for efficient handling of multiple operations.
+pub struct CustomSingleManager {
+    manager: SingleManager,
+    interval_counter: IntervalCounter,
```

Comment:
The idea was that this is counted per connection not for the manager, each connection shall be revalidated after N uses, not after the pool gave out N connections.

Then you also won't need to have this atomic counter thingy, because you have mutable access.